src/core/server/saved_objects/permission_control/acl.ts

```diff
@@ -0,0 +1,249 @@
+      this.permissions[permissionType] = addToPrincipals(
+        this.permissions[permissionType],
+        principals.users,
+        principals.groups
+      );
+    }
+
+    return this;
+  }
+
+  // permissions object build function, remove specific permission of specific principal from the object
+  public removePermission(permissionTypes: string[], principals: Principals) {
+    if (!permissionTypes || !principals) {
+      return this;
+    }
+    if (!this.permissions) {
+      this.permissions = {};
+    }
+
+    for (const permissionType of permissionTypes) {
+      const result = deleteFromPrincipals(
+        this.permissions![permissionType],
+        principals.users,
+        principals.groups
+      );
+      if (result) {
+        this.permissions[permissionType] = result;
+      }
+    }
+
+    return this;
+  }
+
+  /**
+   * transform permissions format
+   * original permissions:   {
+   *     read: {
+   *         users:['user1']
+   *     },
+   *     write:{
+   *         groups:['group1']
+   *     }
+   * }
+   *
+   * transformed permissions: [
+   *     {type:'users',name:'user1',permissions:['read']},
+   *     {type:'groups',name:'group1',permissions:['write']},
+   * ]
+   */
+  public toFlatList(): TransformedPermission[] {
+    const result: TransformedPermission[] = [];
+    if (!this.permissions) {
+      return result;
+    }
+
+    for (const permissionType in this.permissions) {
+      if (Object.prototype.hasOwnProperty.call(this.permissions, permissionType)) {
+        const { users = [], groups = [] } = this.permissions[permissionType] ?? {};
+        users.forEach((user) => {
+          const found = result.find((r) => r.type === PrincipalType.Users && r.name === user);
+          if (found) {
+            found.permissions.push(permissionType);
+          } else {
+            result.push({ type: PrincipalType.Users, name: user, permissions: [permissionType] });
+          }
+        });
+        groups.forEach((group) => {
+          const found = result.find((r) => r.type === PrincipalType.Groups && r.name === group);
+          if (found) {
+            found.permissions.push(permissionType);
+          } else {
+            result.push({ type: PrincipalType.Groups, name: group, permissions: [permissionType] });
+          }
+        });
+      }
+    }
+
+    return result;
+  }
+
+  public resetPermissions() {
+    // reset permissions
+    this.permissions = {};
+  }
+
+  // return the permissions object
+  public getPermissions() {
+    return this.permissions;
+  }
+
+  /**
+   * generate query DSL by the specific conditions, used for fetching saved objects from the saved objects index
+   */
+  public static generateGetPermittedSavedObjectsQueryDSL(
+    permissionTypes: string[],
+    principals: Principals,
+    savedObjectType?: string | string[]
+  ) {
+    if (!principals || !permissionTypes) {
+      return {
+        query: {
+          match_none: {},
```

Comment:
If there is no principals but has permissionTypes, should we return all?

Reply:
I think it's better that the caller skips the function call if there is no principal, this judge is just a parameter check which means no principal shouldn't happen, and for security reason, we should return none here if no principal.